### PR TITLE
C++ example documentation

### PR DESCRIPTION
**Pull request by kalaspuffar**
I was trying to figure out the examples for Bazel and found that the documentation for the first steps was missing so I figured I start the conversation by writing the first copy.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by damienmg** [comment]
Unfortunately I just realized that nobody is watching PR for that repository. sorry for the delay. NYC is offline for the weekend but they should come back at you in a few days.

Sorry again.

File: cpp-tutorial/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Examples to build c++ code
+
+This package will showcase how to build c++ code in stages.
```
Comment (jin) [inline cpp-tutorial/README.md]: nit: C++

File: cpp-tutorial/stage1/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Stage 1
+
+This showcases how to build a single file to create a runnable application.
+
+In the build file for this stage you can read that this will use the c-compiler to build a binary file. (cc_binary)
```
Comment (jin) [inline cpp-tutorial/stage1/README.md]: nit: BUILD file

We can't really infer this from the BUILD file.. here we can say instead "This BUILD file shows that we want to build a C++ binary using the `cc_binary` rule provided by Bazel.`

File: cpp-tutorial/stage1/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+The name of the binary will be hello-world
```
Comment (jin) [inline cpp-tutorial/stage1/README.md]: backticks around `hello-world`

File: cpp-tutorial/stage1/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+And we also name the required source files to build the current application.
```
Comment (jin) [inline cpp-tutorial/stage1/README.md]: "In the `cc_binary` rule, let's name the binary `hello-world`, and provide the required source files in the `srcs` attribute."

File: AUTHORS [hunk 5]
```diff
@@ -8,3 +8,4 @@
 
 Google Inc.
 Manfred Touron <m@42.am>
+Daniel Persson <mailto.woden@gmail.com>
```
Comment (jin) [inline AUTHORS]: please rm

File: cpp-tutorial/README.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+### Stage 1
+The first stage is really simple and shows you how to build a single file.
```
Comment (jin) [inline cpp-tutorial/README.md]: rm "is really simple"

"how to build a single file" -> "how to compile a binary with a single source file"

File: cpp-tutorial/README.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+### Stage 2
+The second stage will showcase how to build an application with multiple files, separated in a library and a binary.
```
Comment (jin) [inline cpp-tutorial/README.md]: s/multiple files/multiple source and header files/

File: cpp-tutorial/README.md [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+### Stage 3
+The third stage showcases how to link multiple build directories building multiple libraries in different packages and then connecting it up with the main application.
```
Comment (jin) [inline cpp-tutorial/README.md]: s/building/by building/

File: cpp-tutorial/stage1/README.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    srcs = ["hello-world.cc"],
+)
+```
+
+To build this example you use (notice that 3 slashes are required in windows)
+```
+bazel build //main:hello-world
+```
+
+The build will output some information for example
```
Comment (jin) [inline cpp-tutorial/stage1/README.md]: If the build is successful, Bazel prints the following output:

File: cpp-tutorial/stage1/README.md [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+```
+cc_binary(
+    name = "hello-world",
+    srcs = ["hello-world.cc"],
+)
+```
+
+To build this example you use (notice that 3 slashes are required in windows)
+```
+bazel build //main:hello-world
```
Comment (jin) [inline cpp-tutorial/stage1/README.md]: Let's also show the Windows example:

```python
bazel build //main:hello-world

# In Windows, note the three slashes

bazel build ///main:hello-world
```

File: cpp-tutorial/stage2/README.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
+# Stage 2
+
+### Library
+
+Below we see a new configuration option. This is a c-compiler library with a seperate header file as an interface, the sources and a name for the library to be built.
```
Comment (jin) [inline cpp-tutorial/stage2/README.md]: `cc_library` is not a configuration option, it's a rule. 

You can say: "Here, we introduce the `cc_library` rule for building C++ libraries. Here, we have a `cc_library` named `hello-greet` and its header and source files are defined accordingly."

File: cpp-tutorial/stage2/README.md [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+cc_library(
+    name = "hello-greet",
+    srcs = ["hello-greet.cc"],
+    hdrs = ["hello-greet.h"],
+)
+```
+
+### Binary
+
+The binary configuration we saw in stage 1 has not changed except that we now depend on the library hello-great.
```
Comment (jin) [inline cpp-tutorial/stage2/README.md]: nit: Stage 1

nit: s/binary/`cc_binary`/

nit: ..has not changed, except that..

nit: s/library/`cc_library`

nit: s/hello-great/`hello-greet`

File: cpp-tutorial/stage2/README.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+cc_binary(
+    name = "hello-world",
+    srcs = ["hello-world.cc"],
+    deps = [
+        ":hello-greet",
+    ],
+)
+```
+
+To build this example you use (notice that 3 slashes are required in windows)
```
Comment (jin) [inline cpp-tutorial/stage2/README.md]: can rm the Windows slashes line, since it's mentioned in stage 1

File: cpp-tutorial/stage3/README.md [hunk 14]
```diff
@@ -0,0 +1,30 @@
+# Stage 3
+
+In this stage we step it up and showcase how to integrate multiple libraries and even in different packages.
```
Comment (jin) [inline cpp-tutorial/stage3/README.md]: "how to integrate multiple `cc_library` targets from different packages"

File: cpp-tutorial/stage3/README.md [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+Below we see a similar configurtion from stage 2. The differance is that this BUILD file is in the lib package / directory. And the new property below visibility will tell bazel which other packages that should be able to see / use this package. So in this case the main package can use this library.
```
Comment (jin) [inline cpp-tutorial/stage3/README.md]: s/configurtion/configuration/

s/differance/difference/

The phrasing sounds a little off here, how about this? "Below, we see a similar configuration from Stage 2, except that this BUILD file is in a subdirectory called `lib`. In Bazel, subdirectories containing BUILD files are known as packages."

The official term for something like `visibility` is an attribute, not property. 

rm "And" from "And the new.."

s/bazel/Bazel

s/main package/`//main` package

replace "see / use" with "reference"

File: cpp-tutorial/stage3/README.md [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+```
+cc_library(
+    name = "hello-time",
+    srcs = ["hello-time.cc"],
+    hdrs = ["hello-time.h"],
+    visibility = ["//main:__pkg__"],
+)
+```
+
+Then to use the library above we add an extra dependency for //[package-path]:[dependency-name].
```
Comment (jin) [inline cpp-tutorial/stage3/README.md]: s/for/in the form of

s/[dependency-name]/[target-name]

File: cpp-tutorial/stage3/README.md [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    name = "hello-world",
+    srcs = ["hello-world.cc"],
+    deps = [
+        ":hello-greet",
+        "//lib:hello-time",
+    ],
+)
+```
+
+To build this example you use (notice that 3 slashes are required in windows)
```
Comment (jin) [inline cpp-tutorial/stage3/README.md]: rm note about the three slashes, since it's mentioned in Stage 1

**Comment by kalaspuffar** [comment]
Hi @jin.

Thank you for the suggested changes. I've looked through them and implemented all as changes to this PR.

I'm not sure if the suggestion to remove the information about the differences in build steps for windows and Linux is a bad thing to have in all the tutorial steps. Sometimes people read them out of order and miss some crucial information. Then again I think it will be apparent when running the build.

Best regards
Daniel

File: cpp-tutorial/stage2/README.md (hunk 11 above)
Comment (jin) [reply cpp-tutorial/stage2/README.md]: Sorry, small fix: can you please remove the "Here," from the second sentence?

**Comment by jin** [comment]
SGTM on leaving the Linux/Windows note in.

I've also sent in my comments on Stage 3, please take a look.

**Comment by kalaspuffar** [comment]
Hi @jin.

I've added the full build instructions back to both Stage 2 and 3 and removed Here, in the second sentence.

Thanks for your time, and please tell me if I've missed or misunderstood something.

Best regards
Daniel

**Comment by jin** [comment]
Thanks for the updates! I think you missed some here: https://github.com/bazelbuild/examples/pull/41/files/66ae6016c793e6f3b2f90f457eb28644b3e4fdec#diff-f3733191e70b5b78a58a0b17a2020f94

**Comment by jin** [comment]
LGTM, @spomorski do you have any inputs?

**Comment by kalaspuffar** [comment]
Hi @jin 

I've missed some backticks at one place and also changed so I use the wording Rule everywhere instead of configuration.

Thank you for your time.

Best regards
Daniel

**Pull request merged**